doc-processor-gradle-plugin/src/main/kotlin/nl/jolanrensen/docProcessor/defaultProcessors/SampleDocProcessor.kt

```diff
@@ -0,0 +1,207 @@
+        )?.getContentBetweenSampleComments()
+            ?: throwError(samplePath, queries)
+
+        // convert the source text to the correct comment content for the current language
+        val commentContent = convertSourceTextToCommentContent(
+            sampleSourceText = targetSourceText,
+            sampleLanguage = targetDocumentable.programmingLanguage,
+            currentLanguage = documentable.programmingLanguage,
+        )
+
+        // add extra content back if it existed
+        return if (extraContent.isNotEmpty()) {
+            buildString {
+                append(commentContent)
+                if (!extraContent.first().isWhitespace())
+                    append(" ")
+                append(extraContent)
+            }
+        } else commentContent
+    }
+
+    private fun throwError(samplePath: String, queries: List<String>): Nothing =
+        error(
+            """
+            |Reference not found: $samplePath. 
+            |Attempted queries: [
+            ${queries.joinToString("\n") { "|  $it" }}
+            ]""".trimMargin()
+        )
+
+    /**
+     * Returns the entire source text of the entire [targetDocumentable], optionally
+     * removing docs at the start with [noComments].
+     * Returns `null` if the source text cannot be found.
+     */
+    private fun getTargetSourceTextOrNull(
+        targetDocumentable: DocumentableWrapper,
+        noComments: Boolean,
+    ): String? {
+        val indent = " ".repeat(targetDocumentable.docIndent)
+        val rawQueriedSource = targetDocumentable.source.psi
+            ?.text
+            ?.let {
+                if (noComments) it.replace(docRegex) { "" }
+                else it
+            } ?: return null
+        return (indent + rawQueriedSource).trimIndent()
+    }
+
+    /**
+     * Trims the source text to only the code between the `// SampleStart` and `// SampleEnd` comments.
+     * If no such comments exist, the entire source text is returned.
+     */
+    private fun String.getContentBetweenSampleComments(): String {
+        val hasSampleComments = sampleStartRegex.containsMatchIn(this) &&
+                sampleEndRegex.containsMatchIn(this)
+
+        return if (hasSampleComments) {
+            val start = sampleStartRegex.findAll(this).first().range.last + 1
+            val end = sampleEndRegex.findAll(this).last().range.first - 1
+            this.substring(start, end).trimIndent()
+        } else this
+    }
+
+    /**
+     * Converts the [sample source text][sampleSourceText] to the correct comment content
+     * for the [sample source text language][sampleLanguage] and
+     * [current language][currentLanguage].
+     */
+    private fun convertSourceTextToCommentContent(
+        sampleSourceText: String,
+        sampleLanguage: ProgrammingLanguage,
+        currentLanguage: ProgrammingLanguage,
+    ): String = buildString {
+        when (currentLanguage) {
+            JAVA -> {
```

Comment:
After one iteration with refactoring to ENUM I want to go deeper and convert it to something with polymorphic calls (sealed classes but not exactly), just thoughts

Reply:
I don't think there are enough differences between the two languages for that. Dokka also doesn't make the distinction per language for a `Documentable`. It's a single common type.